bdk_keychain/src/keychain_txout_index.rs

```diff
@@ -248,6 +250,61 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
         }
     }
 
+    /// Convenience method to call [`store_until_unused_count`] for all keychains.
+    ///
+    /// [`store_until_unused_count`]: Self::store_until_unused_count
+    pub fn store_all_until_unused_count(&mut self, unused_count: u32) -> bool {
+        let keychains = self.keychains.keys().cloned().collect::<Vec<_>>();
+        keychains
+            .into_iter()
+            .map(|keychain| self.store_until_unused_count(&keychain, unused_count))
+            .fold(false, |acc, v| acc || v)
+    }
+
+    /// Derives script pubkeys from the descriptor until the unused script count reaches
+    /// `unused_count`.
+    ///
+    /// Returns whether any new script pubkeys were derived.
+    pub fn store_until_unused_count(&mut self, keychain: &K, unused_count: u32) -> bool {
+        if unused_count < 1 {
+            return false;
+        }
+        let up_to = self
+            .last_active_index(keychain)
+            .map(|i| i.saturating_add(unused_count))
+            .unwrap_or(unused_count.saturating_sub(1));
+        self.store_up_to(keychain, up_to)
+    }
+
+    /// Convenience method to call [`prune_unused`] for all keychains.
+    ///
+    /// `prune_from` contains the lowest allowed index to prune from per keychain (default is 0).
+    ///
+    /// [`prune_unused`]: Self::prune_unused
+    pub fn prune_all_unused(&mut self, prune_from: BTreeMap<K, u32>) -> bool {
+        let keychains = self.keychains.keys().cloned().collect::<Vec<_>>();
+        keychains
+            .into_iter()
+            .map(|keychain| self.prune_unused(&keychain, *prune_from.get(&keychain).unwrap_or(&0)))
+            .fold(false, |acc, v| acc || v)
+    }
+
+    /// Prunes unused script pubkeys of a given keychain, starting from index `prune_from`.
+    ///
+    /// Returns whether anything was removed.
+    pub fn prune_unused(&mut self, keychain: &K, prune_from: u32) -> bool {
```

Comment:
Let's call this `trim_unused` since trim is already used like this in `std`. I think we don't want the `prune_from` parameter. It should just prune to the last active index. The reason you want this is you are trying to restore derivation indices to what they were before right? But we should not be doing this restoration. There are three cases:

1. We are doing an initial scan after restoring from seed words. In this case we don't care about what the derivation indices were before we started because they are all 0.
2. We are updating our wallet with a new block. In this case we don't need to trim since we don't need to do pad with unused since we are the authoritative source on our derivation indices.
3. We are a watch only wallet. In this case we will pad with unused but we never need to trim.

In none of the cases do we need this parameter. This is why I think you need two commands in your rpc example: `Scan` and `Sync` which are scenarios (1) and (2) respectively.

Reply:
Personally I don't think `Sync` is ever a good idea with our model. What if the user has two instances of the same descriptor (`WalletA` and `WalletB`) and the following scenario plays out:
* `WalletA` is currently synced up to height 100 and last_derivation_index is 10.
* `WalletB` receives $$$ via txs that are confirmed in block 101 and last_used_index is 12.
* After some time, `WalletA` initiates a `Sync` up to height 102. Our model will fail to find incoming txs sending to indexes 11 and 12.

Hence, I think it will be safer to always enforce a minimum lookahead count (which is scan).

Reply:
> What if the user has two instances of the same descriptor (`WalletA` and `WalletB`) and the following scenario plays out:

You should not design a system like this because it results in address re-use. I'd say it's a feature if our API makes it really awkward to implement this since it will stop and make you think about other designs. Each system that gives out addresses should be the only system giving out addresses on that keychain. The rule is simple, if you are a system that gives out the addresses do a sync and make sure to manage derivation indices correctly, otherwise do a scan and just pad your keychain with a bunch of pre-derived addresses.